trellis-api/src/main/java/org/trellisldp/api/JoiningResourceService.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.trellisldp.api;
+
+import static java.util.stream.Stream.concat;
+
+import java.time.Instant;
+import java.util.Optional;
+import java.util.concurrent.Future;
+import java.util.stream.Stream;
+
+import org.apache.commons.rdf.api.Dataset;
+import org.apache.commons.rdf.api.IRI;
+import org.apache.commons.rdf.api.Quad;
+
+/**
+ * Uses two underlying persistence services (an {@link ImmutableDataService} and
+ * a {@link MutableDataService}) to provide the {@link ResourceService} API.
+ *
+ * @author ajs6f
+ * @see ImmutableDataService
+ * @see MutableDataService
+ */
+public abstract class JoiningResourceService implements ResourceService {
```

Comment:
Would `AbstractResourceService` be a better name? (or `AbstractResourcePersistenceService` though that's getting pretty long). "Joining" does communicate the connection between mutable and immutable data. There's also no harm in keeping the name as it is and modifying it later when someone finds a better name.

Reply:
I like `AbstractResourceService` better by a bit. But maybe it _is_ better to leave it be for now. If it doesn't really bug you I will leave it be.

Reply:
Feel free to leave this as-is.

Reply:
k, then I'll move on to some tests/examples.